File: 2024/python/day_01.py
#!/usr/bin/env python3

import os.path as op
from collections import Counter

INPUT_FILEPATH = op.join(op.dirname(__file__), "..", "inputs", f"{op.splitext(op.basename(__file__))[0]}.txt")


def parse_input(input: str) -> tuple[list[int], ...]:
    zipped_result: zip[tuple[int, int]] = zip(*[map(int, line.split()) for line in input.splitlines()])
    return tuple(map(list, zipped_result))

def part_1(input: str):
    all_left, all_right = parse_input(input)
    diffs = [abs(l-r) for l, r in zip(sorted(all_left), sorted(all_right))]
    print("Part 1:", sum(diffs))

def part_2(input: str):
    all_left, all_right = parse_input(input)
    right_occurences = Counter(all_right)
    similarities = [v * right_occurences[v] for v in all_left]
    print("Part 2:", sum(similarities))


if __name__ == "__main__":
    with open(INPUT_FILEPATH) as input:
        all_input = input.read()
        part_1(all_input)
        part_2(all_input)

File: 2024/haskell/day_01.hs
#!/usr/bin/env cabal

{- cabal:
build-depends:
  base ^>=4.17.0.0,
  filepath ^>= 1.4.2.2,
-}

{-# LANGUAGE CPP #-}

import System.FilePath
import Data.List (sort)
import Data.Bifunctor (bimap)

inputFile :: FilePath
inputFile = ".." </> "inputs" </> takeFileName __FILE__ -<.> ".txt"

parseInput :: String -> ([Int], [Int])
parseInput input = process (lines input) [] [] where
    process [] left right = (left, right)
    process (x:xs) left right =
      let [l, r] = map read $ words x
      in process xs (l:left) (r:right)

distances :: ([Int], [Int]) -> [Int]
distances (left, right) = zipWith (\x y -> abs (x - y)) left right

similarities :: ([Int], [Int]) -> [Int]
similarities (left, right) = [ l * length (filter (== l) right) | l <- left]

main :: IO ()
main = do
  file <- readFile inputFile
  let parsedInput = parseInput file
      sortedInput = bimap sort sort parsedInput
  putStrLn $ "Part 1: " ++ show (sum $ distances sortedInput)
  putStrLn $ "Part 2: " ++ show (sum $ similarities parsedInput)
